REGRESSION(294049@main): starting an animation via el.animate().reverse only works every other time

graouts opened the pull request:
#### 841997317f49a733a429d2ed52f589a247d762b2
<pre>
REGRESSION(294049@main): starting an animation via el.animate().reverse only works every other time
<a href="https://bugs.webkit.org/show_bug.cgi?id=299823">https://bugs.webkit.org/show_bug.cgi?id=299823</a>
<a href="https://rdar.apple.com/162067522">rdar://162067522</a>

Reviewed by Anne van Kesteren.

We changed the process to mark an animation as &quot;ready&quot; [0] in 294049@main in such a way that we rely
on an enqueued task to process pending animations as a group. The content filed for this bug had a
reversed animation started within an iframe, which may have different timing characteristics than
a more traditional setup since throttling may be involved. As a result, the assumption made by the
code change in 294049@main where animation resolution would happen _after_ pending animations had been
processed and set a pending start time under `AnimationTimelinesController::processPendingAnimations()`
would not necessarily hold true.

We make a simple change that ensures that animation resolution is scheduled, if that is not already the
case, when we set the pending start time of an animation.

We also add a new test which replicates the reported issue where a hosting document has a continuously-
running animation, which ensures animation resolution is scheduled on every rendering frame, and a child
document starts a reversed animation one and again after the first animation&apos;s completion. Since the iframe
will throttle timers, the enqueued task to process pending animations will run only after animation resolution
has been performed as scheduled by the parent document.

[0] <a href="https://drafts.csswg.org/web-animations-1/#ready">https://drafts.csswg.org/web-animations-1/#ready</a>

Test: webanimations/reverse-animation-within-iframe.html
* LayoutTests/webanimations/resources/reverse-animation-iframe.html: Added.
* LayoutTests/webanimations/reverse-animation-within-iframe-expected.txt: Added.
* LayoutTests/webanimations/reverse-animation-within-iframe.html: Added.
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::pendingStartTimeWasSetOnAnimation):
* Source/WebCore/animation/DocumentTimeline.h:
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::setPendingStartTime):
* Source/WebCore/animation/WebAnimation.h:
(WebCore::WebAnimation::setPendingStartTime): Deleted.

Canonical link: <a href="https://commits.webkit.org/301414@main">https://commits.webkit.org/301414@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7dba3ef03cc4b66cefb130ac1d7bb78a209eaddb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125926 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45588 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36349 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132796 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77785 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/029e17b5-2c07-4540-9959-91c8ad2ef7d0) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127797 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46272 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54146 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/95941 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64032 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/41e461b9-f495-4668-8987-445a324549ef) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128874 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37003 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/112618 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76430 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/802104d3-893a-42de-a2ce-e55155d7430f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35909 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30803 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76267 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106785 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31015 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135480 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52711 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40444 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104413 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53160 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108828 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104133 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49505 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27835 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50085 "Hash 7dba3ef0 for PR 52233 does not build (failure)") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19703 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52605 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58417 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51948 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55298 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53647 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/animation/WebAnimation.cpp:
```diff
@@ -1580,6 +1580,14 @@ void WebAnimation::maybeMarkAsReady()
+    m_pendingStartTime = pendingStartTime;
+
+    ASSERT(m_timeline && is<DocumentTimeline>(m_timeline));
+    Ref { *dynamicDowncast<DocumentTimeline>(m_timeline) }->pendingStartTimeWasSetOnAnimation();
```
Why dynamicDowncast here?

graouts replied:
Force of habit, a simple `downcast` will suffice of course.

annevk replied:
That should also allow you to simplify the ASSERT to a non-null check. It might even not be needed at all, but not entirely sure about that.

graouts replied:
I simplified the ASSERT to just `ASSERT(is<DocumentTimeline>(m_timeline))` since the null check was redundant.

I added the assertion here because this code should only be reached with document timelines as the only call site, `AnimationTimelinesController::processPendingAnimations()`, first checks that the animation is attached to a monotonic (ie. non-scroll) timeline and document timelines are the only type of monotonic timeline.

annevk reviewed (approved):
Looks good modulo nit.

graouts committed:
REGRESSION(294049@main): starting an animation via el.animate().reverse only works every other time
https://bugs.webkit.org/show_bug.cgi?id=299823
rdar://162067522

Reviewed by Anne van Kesteren.

We changed the process to mark an animation as "ready" [0] in 294049@main in such a way that we rely
on an enqueued task to process pending animations as a group. The content filed for this bug had a
reversed animation started within an iframe, which may have different timing characteristics than
a more traditional setup since throttling may be involved. As a result, the assumption made by the
code change in 294049@main where animation resolution would happen _after_ pending animations had been
processed and set a pending start time under `AnimationTimelinesController::processPendingAnimations()`
would not necessarily hold true.

We make a simple change that ensures that animation resolution is scheduled, if that is not already the
case, when we set the pending start time of an animation.

We also add a new test which replicates the reported issue where a hosting document has a continuously-
running animation, which ensures animation resolution is scheduled on every rendering frame, and a child
document starts a reversed animation one and again after the first animation's completion. Since the iframe
will throttle timers, the enqueued task to process pending animations will run only after animation resolution
has been performed as scheduled by the parent document.

[0] https://drafts.csswg.org/web-animations-1/#ready

Test: webanimations/reverse-animation-within-iframe.html
* LayoutTests/webanimations/resources/reverse-animation-iframe.html: Added.
* LayoutTests/webanimations/reverse-animation-within-iframe-expected.txt: Added.
* LayoutTests/webanimations/reverse-animation-within-iframe.html: Added.
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::pendingStartTimeWasSetOnAnimation):
* Source/WebCore/animation/DocumentTimeline.h:
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::setPendingStartTime):
* Source/WebCore/animation/WebAnimation.h:
(WebCore::WebAnimation::setPendingStartTime): Deleted.

Canonical link: https://commits.webkit.org/301414@main
